007-OFEP-flag-change-events.md

```diff
@@ -0,0 +1,66 @@
+## 007-OFEP-flag-change-events
+
+## State: DRAFTING
+
+Some flag SDKs support listening for flag value changes or general configuration changes ([launchdarkly](https://docs.launchdarkly.com/sdk/features/flag-changes), [cloudbees](https://docs.cloudbees.com/docs/cloudbees-feature-management/latest/reporting/configuration-fetched-handler), flagd). This can allow us to use an event-based paradigm for consuming flags. Client apps may use feature flags for characteristics that aren't specifically tied to a user-action, making imperative flag evaluation a less-than-ideal solution. A server application could listen to a flag that changes some operational behavior.
+
+Examples:
+
+- A web client's banner color could be updated when an associated flag is changed, without any user action.
+- A web service could subscribe to a flag that controls its log-level.
+
+## Design
+
+The provider interface and the OpenFeature client would be extended to have new functionality to register handlers for a particular flag value or the configuration in general. When a _application author_ registers a handler, the provider would react by appropriately registering a listener on it's SDK, polling it's REST API, etc. If the flag (or configuration in general) is updated, the provider would call the registered handler. The handler might be called with the flag key, or perhaps some other metadata pertaining to the configuration update. **Flag values would not be provided to the handler**, the handler would simply run, indicating the associated configuration had changed. The _application author_ would then perform a flag evaluation for the changed flag. This is consistent with the aforementioned flag systems, which do not evaluate the flags when they've changed. One reason is that no dynamic context can be reasonably provided in the case of events, since the event is driven by a change in the flag management system, not a user-action.
+
+Per-flag design:
+
+```ts
+const client = OpenFeature.getClient();
+
+// subscribe to changes in a string flag called "hex-color", and evaluate it when it's updated
+client.addHandler("hex-color", async (key) => {
+  console.log(
+    `Got update for ${key}, new value is ${await client.getStringValue(
+      key,
+      "000000"
+    )}`
+  );
+});
+```
+
+General configuration change design:
+
+```ts
+const client = OpenFeature.getClient();
+
+// subscribe to changes in the configuration in general, and evaluate flagd it when it's updated
+// this might be easier to implement across vendors, since many don't allow subscription to a particular flag, but general configuration changes
+client.on("configuration-change", async (changeMetadata) => {
```

Comment:
Instead of having both methods, why don't we have changeMetadata just contain a list of flags which were updated if applicable?

Reply:
Do we have an example of `changeMetadata`? I'm only familiar with the LaunchDarkly implementation of flag listeners that can listen for all changes, or a change given a `flag` and a `context`.

Reply:
> Do we have an example of `changeMetadata`?

I don't at this point, it was just food for thought. I somewhat doubt we could get any consistent information across different SDKs that could be useful there.

[Cloudbees has some metadata about the change](https://docs.cloudbees.com/docs/cloudbees-feature-management/latest/reporting/configuration-fetched-handler), but again, it seems like this would be widely different across SDKs, so we might want to leave out any sort of parameter here.

Reply:
I like the ```configuration-change``` approach, but is there a half-way house, were we can define an optional filter.   I'm wondering if you had an application connected to a project with a few hundred flags, but only cared about 10, would you get alot of unnecessary interruptions ?

Reply:
I will add more detail. I didn't want to go too far if there wasn't much interest, but it seems like there is.

Reply:
I've made significant updates to this OFEP, addressing many of the concerns in this thread, hopefully.

Reply:
@weyert I've also updated the change event payload.